_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+Interesującym operatorem jest operator _first()_, który pobiera pierwszy element ze strumienia, następnie jako efekt uboczny - wypisuje obserwatora z listy obserwujących dany strumień! Jest to bardzo wygodne, jeżeli z danego gorącego strumienia potrzebujemy tylko jednej wartości - nie musimy przejmować się w tym momencie wywołaniem funkcji unsubscribe().
+Przykład:
+```typescript
+observable$.pipe(first())
+    .subscribe(value => console.log('pierwsza otrzymana wartosc: ', value))
+// pierwsza otrzymana wartosc: 15000
+```
+### withLatestFrom
+Następnym operatorem wartym uwagi jest _withLatestFrom_, który umożliwia nam 'skrzyżowanie strumieni', czyli dołączenie do jednego strumienia ostatnio emitowanej wartości przez drugi strumień. Przykładowo:
+```typescript
+const otherStream$ = new BehaviourSubject<boolean>(true);
+observable$.pipe(withLatestFrom(otherStream$))
+    .subscribe(([value, otherStreamValue]) => console.log('otrzymana wartosc z pierwszego strumienia: {}, z drugiego: {}', value, otherStreamValue))
+// otrzymana wartosc z pierwszego strumienia: 15000, z drugiego: true
+```
+### takeUntil
+Operator takeUntil jest wygodnym operatorem, jeżeli chcemy w efekciarski i prosty sposób wypisać się z listy obserwujących dany strumień.
+Rzućmy okiem na kod:
+```typescript
+@Component({
+  selector: 'app-some-component',
+  template: `
+        some template
+  `,
+  styleUrls: ['some-styles.scss']
+})
+export class SomeComponent implements OnDestroy, OnInit {
+    private destroySubject$ = new Subject<void>();
+    private someValues$ = new BehaviourSubject<string>('initial value');
+    
+    ngOnInit(): void {
+       this.someValues$.asObservable().pipe(takeUnitl(this.destroySubject$))
+            .subscribe(value => ...)
+    }
+
+    ngOnDestroy(): void {
+       this.destroySubject.next();
+    }   
+
+}
+``` 
+Powyższy fragment przedstawia nam SomeComponent, który nasłuchuje na wartości ze strumienia _someValues$_ aż do momentu, gdy zostanie wyemitowana wartość z drugiego strumienia: _destroySubject$_. W taki oto sposób nie musimy pamiętać o ręcznym odsubskrybowaniu się z pierwszego strumienia, gdyż zostanie on automatycznie zamknięty w momencie zniszczenia komponentu przez Angulara. :)
+### switchMap
+Kolejnym przydatnym operatorem może być _switchMap()_, którego zastosowanie pozwala nam zachować czystość i powstrzyma nas przed tworzeniem tzw. _callback hell_, czyli zagnieżdżania wywołań _subscribe()_ tworzących łańcuch wywołań trudny w czytaniu i utrzymywaniu. Dlatego też zamiast pisać:
+```typescript
+observable$.subscribe(
+    value => someService.processValue(value)
+        .subscribe(
+            someServiceResponse => andYetAnotherService.processAnotherValue(someServiceResponse)
+                .subscribe(...)))
+```
+lepiej zrób:
+```typescript
+observable$.pipe(
+    switchMap(value => someService.processValue(value)),
+    switchMap(someServiceResponse => andYetAnotherService.processAnotherValue(someServiceResponse)))
```

Comment:
hmm generalnie wszystko fajnie, tylko przykład nie praktyczny. Jaką to ma przewagę nam jednym subscribem i wykonaniem tego wszystkiego? :/
Gdzie to się sprawdzi? Bo ten przykład tego nie pokazuje :(

Reply:
opis jest wyżej, nad przykładem :P pozwala uniknac callback hell

Reply:
nie bedzesz mial żadnego callback hell
subscribe(() -> {
const v1 = foo();
const v2 = bar(v1);
const v3 = foobar(v2);
});
sam piszesz, że switchmapa w tym zastosowaniu nie jest tak czytelna. Dlatego chcałbym jakiś przykład, który faktycznie wnosi jakąś wartość

Reply:
dlatego nikt pewnie nie wykorzysta switchMapy bo nadal nie wiadomo, do czego by to zastosować. Bo imo większość zrobi 1 subscribe

Reply:
tu nie chodzi o wywołanie zwykłych funkcji, Twój przykład faktycznie nie tworzy callback hell bo nie operujesz na observablach
subscribe((val) -> {
foo(val).subscribe(v1 => {
       bar(v1).subscribe(v2 => {
                foobar(v2).subscribe(v3 => {
                             // stuff
                       }
        }
};
});
powyzej masz callbackhell

Reply:
teraz rozumie, to proszę ciebie dodaj ten przyklad subscribe hell z opisem odrazu się kmini!